README.md
# jane-puzzles

Solutions to a handful of monthly puzzles posted by Jane Street.

    https://www.janestreet.com/puzzles/

Careful, they're addictive.

## Build

    $ stack build

## Test

    $ stack test


ChangeLog.md
# Changelog for jane-puzzles

## Unreleased changes


test/Spec.hs
import Test.Hspec (shouldBe, it, describe, hspec, xdescribe, xit)

import qualified BobAlice as BA
import qualified WellWell as Well
import qualified GetOutVote as GO

import Data.Array ((!))
import Data.List (foldl', delete)
import Data.Foldable (foldlM)

main :: IO ()
main = hspec $ do
    describe "BobAlice" $ do
        xit "works" $ do
            let r = BA.findNBobWins 3
            map BA.gameTally r `shouldBe` [32, 22, 11]

    describe "GetOutVote" $ do
        it "distributes votes" $ do
            let votes = [("Harry", 15), ("Larry", 30), ("Mary", 57)]
            let dels = GO.distributeDelegates 7 votes
            dels `shouldBe` [("Harry", 1), ("Larry", 2), ("Mary", 4)]

        it "works" $ do
            let expected = [("Harry", 21), ("Larry", 34), ("Mary", 48)]
            fst (head GO.solve) `shouldBe` expected

    describe "WellWell" $ do
        it "prints" $ do
            let s = Well.showGrid Well.initialGrid
            putStrLn s

        it "works" $ do
            done <- foldlM (\prev iter -> do
                let r = Well.drip prev
                putStrLn $ "At iter " ++ show iter ++ "\n" ++ (Well.diffCharWise (Well.showGrid prev) (Well.showGrid r))
                return r) Well.initialGrid [0..359]

            True `shouldBe` True


src/Birthday.hs
{-# LANGUAGE BangPatterns #-}

module Birthday where

import           Data.Array    ((//))
import qualified Data.Array    as Arr
import           Data.List     (foldl', splitAt)
import qualified System.Random as Rand


yearLength :: Int
yearLength = 365

genRandBdays :: (Rand.RandomGen g) => g -> [Int]
genRandBdays = Rand.randomRs (1, yearLength)

canCelebrate :: [Int] -> Bool
canCelebrate bdays
    | length bdays < yearLength = False
    | any (\x -> x == (0 :: Int)) result = False
    | otherwise = True
    where
        bdayMask = Arr.array (1, yearLength) (zip [1..yearLength] (repeat 0))
        result = bdayMask // zip bdays (repeat 1)
        -- result = foldr (\num prev -> prev // [(num, 1)]) bdayMask bdays

likelihood :: (Rand.RandomGen g) => g -> Int -> Int -> Float
likelihood generator iterationCount sampleCount = average
    where
        stuff = genRandBdays generator
        sample = splitAt sampleCount
        (totalCount, successCount) = snd $ foldl' (\(s, (totalCount, successCount)) _ -> do
            let (nums, rest) = sample s
            let !r = canCelebrate nums
            (rest, (totalCount + 1, if r then successCount + 1 else successCount)))
            (stuff, (0, 0))
            [0..iterationCount - 1]
        average = fromIntegral successCount / fromIntegral totalCount


src/BobAlice.hs
module BobAlice where

-- https://www.janestreet.com/puzzles/what-about-bob/
-- https://www.janestreet.com/puzzles/solutions/february-2017-solution/

import           Debug.Trace

data GameState = GameState {
    gamePrevious :: [(Persons, Int)],
    gameTally    :: Int,
    gameDesired  :: Int,
    gameTurns    :: Int
} deriving (Show)

data Persons = Bob | Alice deriving (Show)

initialGame :: Int -> GameState
initialGame desired = GameState {
    gamePrevious = [],
    gameTally = 0,
    gameTurns = 0,
    gameDesired = desired
}

whoWon :: GameState -> Persons
whoWon state = case gameTurns state `mod` 2 of
    -- if game ends in an odd number, alice won
    1 -> Alice
    -- if game ends in an even number, bob won
    0 -> Bob

swapPersons :: Persons -> Persons
swapPersons Alice = Bob
swapPersons Bob = Alice

gameLoop :: GameState -> [GameState]
gameLoop state
    -- shouldn't happen, logic error
    | gameTally state > gameDesired state = error "wat"
    -- yay we won!
    | gameTally state == gameDesired state = [state]
    -- we can't make any choices, we lose :(
    | null options' = [state]
    -- we've lost :/
    | null winningStates = [head newStates]
    -- recurse
    | otherwise = winningStates
    where
        options = case gamePrevious state of
            ((_, num):_) -> filter (/= num) [1..9]
            []           -> [1..9]
        options' = filter (\x -> (x + gameTally state) <= gameDesired state) options
        newStates = concatMap (gameLoop . mkState) options'
        mkState choice = GameState {
            gamePrevious = (swapPersons $ whoWon state, choice) : gamePrevious state,
            gameTally = gameTally state + choice,
            gameTurns = gameTurns state + 1,
            gameDesired = gameDesired state
        }

        -- FIXME(nickhs): do I want the shortest winning states?
        -- find all states that have me (even or odd) as the winner
        isItMe = case gameTurns state `mod` 2 of
            -- I'm even, I win if the final state is odd
            0 -> \num -> num `mod` 2 == 1
            -- I'm odd, I win if the final state is even
            1 -> \num -> num `mod` 2 == 0

        winningStates = filter (isItMe . gameTurns) newStates

findNBobWins :: Int -> [GameState]
findNBobWins count = go 1 []
    where
        go :: Int -> [GameState] -> [GameState]
        go num won = do
            let r = head $ gameLoop (initialGame num)
            let newWon = case whoWon r of
                    Bob -> r : won
                    Alice -> won

            if length newWon == count
            then newWon
            else go (num + 1) newWon


src/GetOutVote.hs
module GetOutVote where

-- https://www.janestreet.com/puzzles/get-out-the-vote/
-- https://www.janestreet.com/puzzles/solutions/june-2016-solution/

import Data.List (sortBy, sortOn, findIndex)
import Data.Either (fromRight)

type Candidate = String

type VotingOutcome = [(Candidate, Int)]
type DelegateOutcome = [(Candidate, Int)]

distributeDelegates :: Int -> VotingOutcome -> DelegateOutcome
distributeDelegates delegateCount votes = do
    let totalVotes = sum $ map snd votes
    let delegates voteCount = fromIntegral delegateCount * (fromIntegral voteCount / fromIntegral totalVotes)
    let delegates' voteCount = floor (delegates voteCount)
    let remainder voteCount = delegates voteCount - fromIntegral (delegates' voteCount)
    let r = sortBy (\(_, _, a) (_, _, b) -> b `compare` a)
            $ map (\(cand, voteCount) ->
                (cand, delegates' voteCount, remainder voteCount)) votes
    let remainingDelegates = delegateCount - sum (map (\(_, x, _) -> x) r)
    let delegateMask = replicate remainingDelegates 1 ++ repeat 0
    let r2 = zipWith (\(cand, delCount, _) moreDels ->
                        (cand, delCount + moreDels)) r delegateMask
    sortOn fst r2

votingOutcomes :: Int -> [VotingOutcome]
votingOutcomes votes = map (\(h, l, m) -> [("Harry", h), ("Larry", l), ("Mary", m)]) validVotes
    where
        validVotes = filter (\(a, b, c) -> a + b + c == votes) allVotes
        allVotes = [(a, b, c) | a <- [0..votes], b <- [0..votes], c <- [0..votes]]

findMatchingVotes :: DelegateOutcome -> Int -> [(VotingOutcome, DelegateOutcome)]
findMatchingVotes delegates voteCount = do
    let vOutcomes = votingOutcomes voteCount
    let delegateCount = sum $ map snd delegates
    let delegateOutcomes = zip vOutcomes $ map (distributeDelegates delegateCount) vOutcomes
    filter (\(votes, dgates) -> dgates == delegates) delegateOutcomes

addVotes :: Candidate -> Int -> VotingOutcome -> Either VotingOutcome VotingOutcome
addVotes cand votes state = case findIndex (\(c, _) -> c == cand) state of
                                Nothing -> Left state
                                Just idx -> Right $ do
                                    let count = snd (state !! idx)
                                    let (before, after) = splitAt idx state
                                    before ++ [(cand, count + votes)] ++ tail after

solve :: [(DelegateOutcome, VotingOutcome)]
solve = filter (\(_, [("Harry", n), _, _]) -> n == 1) newDistributions
    where
        firstKnown = [("Harry", 2), ("Larry", 2), ("Mary", 3)]
        votesMatchingFirst = map fst $ findMatchingVotes firstKnown 102
        newPotentialVotes = map (fromRight (error "wat") . addVotes "Harry" 1) votesMatchingFirst
        newDistributions = zip newPotentialVotes $ map (distributeDelegates 8) newPotentialVotes


src/WellWell.hs
{-# LANGUAGE ScopedTypeVariables #-}

module WellWell where

-- https://www.janestreet.com/puzzles/well-well-well/

import Data.Array (Array, (//), (!))
import qualified Data.Array as Arr
import Data.Text (Text)
import qualified Data.Text as T
import Data.List (intersperse, intercalate, foldl')
import Text.Printf

import Debug.Trace

type Coord = (Int, Int)
data Tile = Tile Int Float deriving (Show, Eq)

-- zero index from bottom left
type Grid = Array Coord Tile

rowLen :: Int
rowLen = 7

colLen :: Int
colLen = 7

bounds :: (Coord, Coord)
bounds = ((0, 0), (rowLen - 1, colLen - 1))

crossProduct :: [a] -> [b] -> [(b, a)]
crossProduct as bs = [(b, a) | a <- as, b <- bs]

inp :: [Int]
inp = [
  01, 05, 27, 22, 28, 40, 14,
  39, 13, 17, 30, 41, 12, 02,
  32, 35, 24, 25, 19, 47, 34,
  16, 33, 10, 42, 07, 44, 18,
  03, 08, 45, 37, 04, 21, 20,
  15, 46, 38, 06, 26, 48, 49,
  09, 23, 31, 29, 11, 36, 43]

allCoords :: [Coord]
allCoords = crossProduct [0..colLen - 1] [0..rowLen - 1]

initialGrid :: Grid
initialGrid = Arr.array bounds tiles
    where
        tiles = zip allCoords (map (\x -> Tile x 0) inp)

orthAdjNeighbours :: ([Coord] -> [Coord]) -> Coord -> [Coord]
orthAdjNeighbours clipper (x, y) = clipper [left, right, top, bottom]
    where
        left = (x - 1, y)
        right = (x + 1, y)
        top = (x, y - 1)
        bottom = (x, y + 1)

clip :: (Coord, Coord) -> [Coord] -> [Coord]
clip ((minX, minY), (maxX, maxY)) = filter isInBounds
    where
        isInBounds (x, y) = isInX x && isInY y
        isInX x = minX <= x && x <= maxX
        isInY y = minY <= y && y <= maxY

wellNeighbours :: Coord -> [Coord]
wellNeighbours = orthAdjNeighbours (clip bounds)

{--
   Function that searches for any water that should be flowing
   and then moves it accordingly
--}
flow' :: Grid -> Grid
flow' grid = foldl' tileFlow grid allCoords

{-- Moves water from tile to neighbours.
    Determines if there's a gradient of water to move, and moves it
--}
tileFlow :: Grid -> Coord -> Grid
tileFlow grid coord = do
    let (Tile tMaxDepth tCurDepth) = grid ! coord
    let neighbours = wellNeighbours coord
    let neighbours' = map (\c -> (c, grid ! c)) neighbours
    let flowers = filter (\(c, Tile maxDepth curDepth) -> (fromIntegral maxDepth - curDepth) > (fromIntegral tMaxDepth - tCurDepth)) neighbours'

    if null flowers
    then grid
    else do
        let minDelta = minimum $ map (\(_, Tile maxDepth curDepth) -> (fromIntegral maxDepth - curDepth) - (fromIntegral tMaxDepth - tCurDepth)) flowers
        let waterSize = min minDelta tCurDepth
        let flowRate = waterSize / fromIntegral (length flowers)
        let newCells = map (\(c, Tile maxDepth curDepth) -> if curDepth + flowRate > fromIntegral maxDepth then error "wat ??" else (c, Tile maxDepth (curDepth + flowRate))) flowers
        let newGrid = grid // (((coord, Tile tMaxDepth (tCurDepth - waterSize))) : newCells)
        newGrid
        -- trace (showGrid newGrid) newGrid

flowUntilStable :: Grid -> Int -> Grid
flowUntilStable g count
  | count > 1000 = g
  | newG == g = g
  | otherwise = flowUntilStable newG (count + 1)
  where
      newG = flow' g
      newG' = trace (diffCharWise (showGrid newG) (showGrid g)) newG

drip :: Grid -> Grid
drip grid = flowUntilStable (grid // [((0, 0), Tile 1 1)]) 0

showGrid :: Grid -> String
showGrid g = do
    let cellWidth = 6 :: Int -- for curDepth maxDepth
    let cellHeight = 1

    let lineBreak = concatMap (\_ -> " " ++ concat (replicate (cellWidth - 1) "-")) [0..colLen - 1]
    let mkRow y = concatMap (\x -> do
            let (Tile maxDepth curDepth) = g ! (x, y)
            printf "|%5.2f" (fromIntegral maxDepth - curDepth)) [0..colLen - 1]
    let rows = concat $ intersperse ("\n" ++ lineBreak ++ "\n") $ map mkRow [0..rowLen - 1] :: String
    concat [lineBreak, "\n", rows, "\n"]

diffCharWise :: String -> String -> String
diffCharWise a b = go 0 False
    where
        defaultColorEscapeCode = "\x1b[0m"
        highlightColorCode = "\x1b[2m"
        isCharDiff idx = a !! idx == b !! idx
        go :: Int -> Bool -> String
        go idx isDiff = if idx >= length a
                        then ""
                        else
                            case (isDiff, isCharDiff idx) of
                                (True, False) -> defaultColorEscapeCode ++ [b !! idx] ++ go (idx + 1) False
                                (False, True) -> highlightColorCode ++ [b !! idx] ++ go (idx + 1) True
                                (False, False) -> [b !! idx] ++ go (idx + 1) isDiff
                                (True, True) -> [b !! idx] ++ go (idx + 1) isDiff
